### PR TITLE
ubuntu24.04 precompile image support

### PR DESCRIPTION
**Pull request by shivakunv**

### CODE REVIEW

**Comment by cdesiniotis** [comment]
@shivakunv https://github.com/NVIDIA/gpu-driver-container/pull/124 has been merged. Please rebase off the latest main branch.

File: .github/workflows/precompiled.yaml [hunk 1]
```diff
@@ -15,9 +15,20 @@
...
 name: Precompiled images
 
+# on:
+#   schedule:
+#     - cron: '00 09 * * *' # scheduled job
+
 on:
-  schedule:
-    - cron: '00 09 * * *'
+  pull_request:
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:23]: for testing

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -40,7 +40,7 @@ jobs:
         aws_access_key_id: ${{ secrets.AWS_ACCESS_KEY_ID }}
         aws_secret_access_key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
         aws_ssh_key: ${{ secrets.AWS_SSH_KEY }}
-        holodeck_config: "tests/holodeck.yaml"
+        holodeck_config: "tests/holodeck_ubuntu22.04.yaml"
```
Comment (shivakunv) [inline .github/workflows/ci.yaml:43]: in ci file we run e2e tests on ubuntu22.04 only

File: .github/workflows/precompiled.yaml [hunk 3]
```diff
@@ -222,6 +269,15 @@ jobs:
           DRIVER_BRANCHES=($(echo "$driver_branch_json" | jq -r '.[]'))
           echo "DRIVER_BRANCHES=${DRIVER_BRANCHES[*]}" >> $GITHUB_ENV
 
+      # step added to skip azure e2e tests and publish the image
+      # FIXME -- remove step once azure kernel upgrade starts working
+      # - name: Skip azure e2e
+      #   run: |
+      #     if [[ "${KERNEL_VERSION}" == *-azure ]]; then
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:276]: for testing

File: .github/workflows/image.yaml [hunk 4]
```diff
@@ -21,12 +21,12 @@ on:
       - opened
       - synchronize
     branches:
-      - main
-      - release-*
+      - maini-no
```
Comment (shivakunv) [inline .github/workflows/image.yaml:24]: for testing

File: .github/workflows/precompiled.yaml [hunk 5]
```diff
@@ -15,9 +15,20 @@
 # Run this workflow on a schedule
 name: Precompiled images
 
+# on:
+#   schedule:
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:19]: for testing

File: .nvidia-ci.yml [hunk 6]
```diff
@@ -187,7 +211,7 @@ image-rhel8:
 .scan-ubuntu24.04:
   # Repeat for each DRIVER_VERSION
   extends:
-    - .driver-versions-ubuntu24.04
+    - .driver-versions
```
Comment (cdesiniotis) [inline .nvidia-ci.yml:214]: Why was this change made?

File: .nvidia-ci.yml [hunk 7]
```diff
@@ -306,6 +342,25 @@ release:ngc-ubuntu22.04:
     - .dist-ubuntu22.04
     - .driver-versions
 
+release:ngc-ubuntu24.04:
+  extends:
+    - .release:ngc
+    - .dist-ubuntu24.04
+    - .driver-versions
```
Comment (cdesiniotis) [inline .nvidia-ci.yml:349]: Note, we excluded this for the time being since we do not want to enable publishing of `ubuntu24.04` tags yet. Can we remove this and `release:ngc-precompiled-ubuntu24.04` until we have some QA coverage on these images? Let's just comment them out so we don't lose this work.

File: .nvidia-ci.yml [hunk 8]
```diff
@@ -455,6 +527,9 @@ sign:ngc-ubuntu-rhel-rhcos:
     - .sign:ngc
   parallel:
     matrix:
+    - SIGN_JOB_NAME: ["ubuntu"]
+      VERSION: ["24.04"]
+      DRIVER_VERSION: ["535.216.03", "550.127.08"]
```
Comment (cdesiniotis) [inline .nvidia-ci.yml:532]: I believe we can remove 535. I don't see Ubuntu 24.04 as a supported distribution in the latest R535 driver support matrix: https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-216-03/index.html#hardware-software-support

File: Makefile [hunk 9]
```diff
@@ -54,9 +54,9 @@ OUT_IMAGE_TAG = $(OUT_IMAGE_VERSION)-$(OUT_DIST)
 OUT_IMAGE = $(OUT_IMAGE_NAME):$(OUT_IMAGE_TAG)
 
 ##### Public rules #####
-DISTRIBUTIONS := ubuntu18.04 ubuntu20.04 ubuntu22.04 ubuntu24.04 signed_ubuntu20.04 signed_ubuntu22.04 signed_ubuntu24.04 rhel8 rhel9 flatcar fedora36 sles15.3 precompiled_rhcos
+DISTRIBUTIONS := ubuntu18.04 ubuntu20.04 ubuntu22.04 ubuntu24.04 signed_ubuntu24.04 signed_ubuntu20.04 signed_ubuntu22.04 rhel8 rhel9 flatcar fedora36 sles15.3 precompiled_rhcos
 PUSH_TARGETS := $(patsubst %, push-%, $(DISTRIBUTIONS))
-BASE_FROM := jammy focal
+BASE_FROM := jammy focal noble
```
Comment (cdesiniotis) [inline Makefile:59]: nit: let's preserve the descending order from left to right.
```suggestion
BASE_FROM := noble jammy focal
```

File: Makefile [hunk 10]
```diff
@@ -54,9 +54,9 @@ OUT_IMAGE_TAG = $(OUT_IMAGE_VERSION)-$(OUT_DIST)
 OUT_IMAGE = $(OUT_IMAGE_NAME):$(OUT_IMAGE_TAG)
 
 ##### Public rules #####
-DISTRIBUTIONS := ubuntu18.04 ubuntu20.04 ubuntu22.04 ubuntu24.04 signed_ubuntu20.04 signed_ubuntu22.04 signed_ubuntu24.04 rhel8 rhel9 flatcar fedora36 sles15.3 precompiled_rhcos
+DISTRIBUTIONS := ubuntu18.04 ubuntu20.04 ubuntu22.04 ubuntu24.04 signed_ubuntu24.04 signed_ubuntu20.04 signed_ubuntu22.04 rhel8 rhel9 flatcar fedora36 sles15.3 precompiled_rhcos
```
Comment (cdesiniotis) [inline Makefile:57]: nit: 
```suggestion
DISTRIBUTIONS := ubuntu18.04 ubuntu20.04 ubuntu22.04 ubuntu24.04 signed_ubuntu20.04 signed_ubuntu22.04 signed_ubuntu24.04 rhel8 rhel9 flatcar fedora36 sles15.3 precompiled_rhcos
```

File: tests/scripts/findkernelversion.sh [hunk 11]
```diff
@@ -17,8 +17,28 @@ chmod a+x bin/regctl
...
+        if [[ $artifact == $prefix*-$suffix ]]; then
+            gh run download --name "$artifact" --dir ./
+            tar -xf $artifact.tar 
+            rm -f $artifact.tar
+            export $(grep -oP 'KERNEL_VERSION=[^ ]+' ./kernel_version.txt)
+            rm -f kernel_version.txt
+            break
+        fi
+    done
+fi
```
Comment (cdesiniotis) [inline tests/scripts/findkernelversion.sh:41]: Question -- why was this added?

File: .github/workflows/precompiled.yaml [hunk 12]
```diff
@@ -70,6 +86,14 @@ jobs:
           echo "PUSH_ON_BUILD=${GENERATE_ARTIFACTS}" >> $GITHUB_ENV
           echo "BUILD_MULTI_ARCH_IMAGES=${GENERATE_ARTIFACTS}" >> $GITHUB_ENV
 
+          if [ "$DIST" == "ubuntu22.04" ]; then
+            echo  "BASE_TARGET=jammy" >> $GITHUB_OUTPUT
+            echo "LTS_KERNEL=5.15" >> $GITHUB_OUTPUT
+          elif [ "$DIST" == "ubuntu24.04" ]; then
+            echo  "BASE_TARGET=noble" >> $GITHUB_OUTPUT
+            echo "LTS_KERNEL=6.8" >> $GITHUB_OUTPUT
+          fi
```
Comment (cdesiniotis) [inline .github/workflows/precompiled.yaml:95]: To avoid duplicating this, what about pushing this logic to the Makefile / Dockerfile? That way we do not need to set these variables when running `make build-base-${BASE_TARGET}`

In the Makefile, we can define the `BASE_TARGET` variable for the respective distro targets:
```
build-signed_ubuntu22.04%: BASE_TARGET = jammy

build-signed_ubuntu24.04%: BASE_TARGET = noble
``` 

For `LTS_KERNEL`, we can set a default value for this ARG in the Dockerfile so that we do not need to set anything here. For example, for the jammy base image, we would have:
```
ARG LTS_KERNEL=5.15
ENV LTS_KERNEL=${LTS_KERNEL}
```

File: .common-ci.yml [hunk 13]
```diff
@@ -88,6 +88,13 @@ trigger-pipeline:
       - DRIVER_BRANCH: [535, 550]
         KERNEL_FLAVOR: [aws, azure, generic, nvidia, oracle]
 
+# Define the matrix of precompiled jobs that can be run in parallel for ubuntu24.04
+.driver-versions-precompiled-ubuntu24.04:
+  parallel:
+    matrix:
+      - DRIVER_BRANCH: [535, 550]
```
[the remark this answers is not in the document]
Comment (shivakunv) [reply .common-ci.yml:95]: build and publish have condition : 
exclude:
          - dist: ubuntu24.04
            driver_branch: 535     
        
        e2e-test has condition: 
        
if [ "$DIST" == "ubuntu24.04" ]; then
            DRIVER_BRANCHES=($(for branch in "${DRIVER_BRANCHES[@]}"; do
            [[ $branch != "535" ]] && echo "$branch"
          done))

File: tests/scripts/findkernelversion.sh (hunk 11 above)
Comment (shivakunv) [reply tests/scripts/findkernelversion.sh:41]: added for first release of ubuntu24.04 or other distro

File: .common-ci.yml [hunk 14]
```diff
@@ -172,6 +179,14 @@ trigger-pipeline:
...
     - if: $CI_PIPELINE_SOURCE != "schedule"
 
+.release-ubuntu24.04:
+  # Perform for each DRIVER_VERSION
+  extends:
+    - .release-generic
+    - .driver-versions
+  rules:
+    - if: $CI_PIPELINE_SOURCE != "schedule"
+
```
[the remark this answers is not in the document]
Comment (shivakunv) [reply .common-ci.yml:189]: pushed by mistake . removed

File: .nvidia-ci.yml (hunk 6 above)
Comment (shivakunv) [reply .nvidia-ci.yml:214]: corrected

File: .nvidia-ci.yml (hunk 8 above)
Comment (shivakunv) [reply .nvidia-ci.yml:532]: corrected

File: tests/scripts/findkernelversion.sh [hunk 15]
```diff
@@ -28,3 +48,5 @@ if [[ $status -eq 0 ]]; then
 else
     export should_continue=true
 fi
+#SHIVA
+export should_continue=true
```
Comment (shivakunv) [inline tests/scripts/findkernelversion.sh:52]: for testing

File: .github/workflows/precompiled.yaml (hunk 12 above)
Comment (shivakunv) [reply .github/workflows/precompiled.yaml]: moved LTS_KERNEL and BASE_TARGET   to set-driver-version-matrix   ( generic solution)
 
 `  # BASE_TARGET setup
          declare -A MAP_BASE_TARGETS
          MAP_BASE_TARGETS["ubuntu22.04"]="jammy"
          MAP_BASE_TARGETS["ubuntu24.04"]="noble"
          BASE_TARGETS_JSON="{"
          for key in "${!MAP_BASE_TARGETS[@]}"; do
            BASE_TARGETS_JSON+="\"$key\":\"${MAP_BASE_TARGETS[$key]}\","
          done
          BASE_TARGETS_JSON="${BASE_TARGETS_JSON%,}}"
          echo "BASE_TARGETS=$BASE_TARGETS_JSON" >> "$GITHUB_OUTPUT"

          # LTS_KERNELS env setup
          declare -A MAP_LTS_KERNELS
          MAP_LTS_KERNELS["ubuntu22.04"]="5.15"
          MAP_LTS_KERNELS["ubuntu24.04"]="6.8"
          LTS_KERNELS_JSON="{"
          for key in "${!MAP_LTS_KERNELS[@]}"; do
            LTS_KERNELS_JSON+="\"$key\":\"${MAP_LTS_KERNELS[$key]}\","
          done
          LTS_KERNELS_JSON="${LTS_KERNELS_JSON%,}}"
          echo "LTS_KERNELS=$LTS_KERNELS_JSON" >> "$GITHUB_OUTPUT"`
          
          extraction 
          
           ```
 BASE_TARGETS_JSON='${{ needs.set-driver-version-matrix.outputs.BASE_TARGETS }}'
          BASE_TARGET=$(echo "$BASE_TARGETS_JSON" | jq -r --arg dist "$DIST" '.[$dist]')
          echo "BASE_TARGET=$BASE_TARGET" >> $GITHUB_ENV
          LTS_KERNELS_JSON='${{ needs.set-driver-version-matrix.outputs.LTS_KERNELS }}'
          LTS_KERNEL=$(echo "$LTS_KERNELS_JSON" | jq -r --arg dist "$DIST" '.[$dist]')
          echo "LTS_KERNEL=$LTS_KERNEL" >> $GITHUB_ENV
          
          
            BASE_TARGET: ${{ env.BASE_TARGET }}
          LTS_KERNEL: ${{ env.LTS_KERNEL }}
```

moved as we are using it at multiple : make and finding kernel version also . we need to change it all those places . 

make DRIVER_BRANCH=${{ matrix.driver_branch }} KERNEL_FLAVOR=${{ matrix.flavor }} LTS_KERNEL=${LTS_KERNEL} build-base-${BASE_TARGET}

          trap "docker rm -f base-${BASE_TARGET}-${{ matrix.flavor }}" EXIT
          docker run -d --name base-${BASE_TARGET}-${{ matrix.flavor }} ghcr.io/nvidia/driver:base-${BASE_TARGET}-${{ matrix.flavor }}-${{ matrix.driver_branch }}
          
          findkernelversion.sh also using it

File: .nvidia-ci.yml (hunk 7 above)
Comment (shivakunv) [reply .nvidia-ci.yml]: corrected

File: Makefile (hunk 9 above)
Comment (shivakunv) [reply Makefile]: corrected

File: Makefile (hunk 10 above)
Comment (shivakunv) [reply Makefile]: corrected

File: .common-ci.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (cdesiniotis) [reply .common-ci.yml:95]: The condition you mention is in the GitHub Action workflows. The current file is the Gitlab CI definition.

[the remark this answers is not in the document]
Comment (shivakunv) [reply .common-ci.yml:95]: Will correct this. comment is for precompiled.yaml

File: .github/workflows/precompiled.yaml (hunk 12 above)
Comment (shivakunv) [reply .github/workflows/precompiled.yaml]: TODO:- added for next PR

Comment (shivakunv) [reply .github/workflows/precompiled.yaml]: @cdesiniotis 

removed map based BASE_TARGET and LTS_KERNEL and followed much cleaner approach  :
definition: 
      base_target_ubuntu22_04: ${{ steps.extract_driver_branch.outputs.base_target_ubuntu22_04 }}
      lts_kernel_ubuntu22_04: ${{ steps.extract_driver_branch.outputs.lts_kernel_ubuntu22_04 }}
      BASE_TARGET_ubuntu24_04: ${{ steps.extract_driver_branch.outputs.base_target_ubuntu24_04 }}
      lts_kernel_ubuntu24_04: ${{ steps.extract_driver_branch.outputs.lts_kernel_ubuntu24_04 }} 

set : 
   # BASE_TARGET and LTS_KERNEL setup
   echo "base_target_ubuntu22_04=jammy" >> $GITHUB_OUTPUT
   echo "base_target_ubuntu24_04=noble" >> $GITHUB_OUTPUT
   echo "lts_kernel_ubuntu22_04=5.15" >> $GITHUB_OUTPUT
   echo "lts_kernel_ubuntu24_04=6.8" >> $GITHUB_OUTPUT

extraction: 
  VAR_NAME_BASE_TARGET=$(echo BASE_TARGET_${DIST} | sed 's/\./_/g')
  echo "VAR_NAME_BASE_TARGET=${VAR_NAME_BASE_TARGET}" >> $GITHUB_ENV
  VAR_NAME_LTS_KERNEL=$(echo LTS_KERNEL_${DIST} | sed 's/\./_/g')
  echo "VAR_NAME_LTS_KERNEL=${VAR_NAME_LTS_KERNEL}" >> $GITHUB_ENV


    env:
          BASE_TARGET: ${{ needs.set-driver-version-matrix.outputs[env.VAR_NAME_BASE_TARGET] }}
          LTS_KERNEL: ${{ needs.set-driver-version-matrix.outputs[env.VAR_NAME_LTS_KERNEL] }}

File: tests/scripts/ci-precompiled-helpers.sh [hunk 16]
```diff
@@ -11,9 +11,10 @@ get_kernel_versions_to_test() {
 
     kernel_versions=()
     for kernel_flavor in "${KERNEL_FLAVORS[@]}"; do
-        # FIXME -- remove if condition, once azure kernel upgrade starts working
-        if [[ "$kernel_flavor" == "azure" ]]; then
-            continue
+        # TODO remove this check once nvidia avaialble
+        # currently for ubuntu24.04 flavor = nvidia-lowlatency
+        if [[ "$DIST" == "ubuntu24.04" && "$kernel_flavor" == "nvidia" ]]; then
```
Comment (shivakunv) [inline tests/scripts/ci-precompiled-helpers.sh:16]: @cdesiniotis please check , currently kernel-flavor = nvidia-lowlatency

File: tests/scripts/findkernelversion.sh [hunk 17]
```diff
@@ -17,8 +17,31 @@ chmod a+x bin/regctl
 export PATH=$(pwd)/bin:${PATH}
 
 # calculate kernel version of latest image
-regctl image get-file ghcr.io/nvidia/driver:base-${BASE_TARGET}-${KERNEL_FLAVOR}-${DRIVER_BRANCH} /var/kernel_version.txt ./kernel_version.txt
-export $(grep -oP 'KERNEL_VERSION=[^ ]+' ./kernel_version.txt)
+regctl image get-file ghcr.io/nvidia/driver:base-${BASE_TARGET}-${KERNEL_FLAVOR}-${DRIVER_BRANCH} /var/kernel_version.txt ./kernel_version.txt 2>/dev/null || true
+if [[ -f ./kernel_version.txt && -s ./kernel_version.txt ]]; then
```
Comment (cdesiniotis) [inline tests/scripts/findkernelversion.sh:21]: As discussed offline, let's see if we can remove this conditional logic as we are building the base image in a previous step in this workflow.

File: .github/workflows/precompiled.yaml [hunk 18]
```diff
@@ -146,42 +171,105 @@ jobs:
...
           if [ -z "$KERNEL_VERSIONS" ]; then
             # no new kernel release
             echo "Skipping e2e tests"
             exit 0
           fi
+          # Remove duplicates
+          KERNEL_VERSIONS=($(printf "%s\n" "${KERNEL_VERSIONS[@]}" | sort -u))
+          for i in "${!KERNEL_VERSIONS[@]}"; do
+            KERNEL_VERSIONS[$i]="${KERNEL_VERSIONS[$i]}-$DIST"
+          done
```
[the remark this answers is not in the document]
Comment (shivakunv) [reply .github/workflows/precompiled.yaml]: yes. removed DIST dependency . 
moved this to ci-helper

File: tests/scripts/findkernelversion.sh [hunk 19]
```diff
@@ -17,8 +17,31 @@ chmod a+x bin/regctl
...
+    # Define variables for artifact pattern
+    prefix="kernel-version-${DRIVER_BRANCH}-${LTS_KERNEL}"
+    suffix="${kernel_flavor}-${DIST}"
+    suffixlowlatency="${kernel_flavor}-lowlatency-${DIST}"
+    artifacts=$(gh api -X GET /repos/${GITHUB_REPOSITORY}/actions/runs/${GITHUB_RUN_ID}/artifacts --jq '.artifacts[].name')
+    # Use a loop or a pattern to find the matching artifact dynamically
+    for artifact in $artifacts; do
+        # TODO remove this check once nvidia avaialble
+        # currently for ubuntu24.04 kernel_flavor = nvidia-lowlatency
+        if [[ $artifact == $prefix*-$suffix || $artifact == $prefix*-$suffixlowlatency ]]; then
```
[the remark this answers is not in the document]
Comment (shivakunv) [reply tests/scripts/findkernelversion.sh]: updated

File: tests/scripts/findkernelversion.sh (hunk 17 above)
Comment (shivakunv) [reply tests/scripts/findkernelversion.sh:21]: this require pushing of base image. 

TODO: follow up added

**Comment by shivakunv** [comment]
skip_remaining  removed for azure and exclude_matrix_values  for azure kernel version added

File: .nvidia-ci.yml [hunk 20]
```diff
@@ -433,6 +489,23 @@ release:ngc-rhel8.10:
...
+  variables:
+    DIST: signed_ubuntu24.04
+    BASE_TARGET: noble
+    PRECOMPILED: "true"
+  needs:
+    - release:ngc-precompiled-ubuntu24.04
+  rules:
+    # Only run NGC release job on scheduled pipelines
+    - if: $CI_PIPELINE_SOURCE == "schedule"
+
```
Comment (cdesiniotis) [inline .nvidia-ci.yml:508]: @shivakunv this needs to be commented out as well since it is referring to release:ngc-precompiled-ubuntu24.04 which is commented out.

```suggestion
#sign:ngc-precompiled-ubuntu24.04:
#  extends:
#    - .driver-versions-precompiled-ubuntu24.04
#    - .dist-ubuntu22.04
#    - .release-generic
#    - .release:ngc-variables
#    - .sign:ngc
#  variables:
#    DIST: signed_ubuntu24.04
#    BASE_TARGET: noble
#    PRECOMPILED: "true"
#  needs:
#    - release:ngc-precompiled-ubuntu24.04
#  rules:
#    # Only run NGC release job on scheduled pipelines
#    - if: $CI_PIPELINE_SOURCE == "schedule"

```

File: .nvidia-ci.yml [hunk 21]
```diff
@@ -455,6 +528,9 @@ sign:ngc-ubuntu-rhel-rhcos:
     - .sign:ngc
   parallel:
     matrix:
+    - SIGN_JOB_NAME: ["ubuntu"]
+      VERSION: ["24.04"]
+      DRIVER_VERSION: ["550.127.08"]
```
Comment (cdesiniotis) [inline .nvidia-ci.yml:533]: Let's comment this out as well.

```suggestion
#    - SIGN_JOB_NAME: ["ubuntu"]
#      VERSION: ["24.04"]
#      DRIVER_VERSION: ["550.127.08"]
```

**Review by cdesiniotis — APPROVED** [review]
